Fix mypy issues in tests and examples

dantp-ai opened the pull request:
Closes #952 

- `SamplingConfig` supports `batch_size=None`. #1077
- tests and examples are covered by `mypy`. #1077
- `NetBase` is more used, stricter typing by making it generic. #1077
- `utils.net.common.Recurrent` now receives and returns a `RecurrentStateBatch` instead of a dict. #1077

dantp-ai committed:
Remove mutable logger_factory used by examples

 * Creating a new instance of LoggerFactoryDefault within each module

dantp-ai committed:
Fix mypy issues

dantp-ai committed:
Add type annotations to funcs signatures

  * Remove func `compute_reward_fn` as it was defined already before

dantp-ai committed:
Rename variable to resolve type conflict

  * Same variable name used multiple times for objects of different types

dantp-ai committed:
Provide right input type to buffer methods

dantp-ai committed:
Index sample from buffer according to cheatsheet recommendations

dantp-ai committed:
Type index to make mypy happy

  * index can be int or np.ndarray. since it's defined twice with the same name, mypy gets confused

dantp-ai committed:
Make mypy happy and check for mask attribute before asserting tests

dantp-ai committed:
Make mypy happy and specify union of types

dantp-ai committed:
Make mypy happy and type ndarray

dantp-ai committed:
Use recommended outer buffer indexing

dantp-ai committed:
Access buffer attrs with __getattr__

dantp-ai committed:
Ignore mypy issue as it explicitly tests for invalid type

dantp-ai committed:
Make mypy happy & use explicit var typing and ignore

dantp-ai committed:
Ignore type on explicit error

dantp-ai committed:
Remove redundant assert

dantp-ai committed:
Merge branch 'master' into refactoring/mypy-issues-test-and-examples

dantp-ai committed:
Add typing to func args

dantp-ai committed:
Add type annotations to funcs

dantp-ai committed:
Fix mypy issues

dantp-ai committed:
Use DataclassPPrintMixin to print collector stats

dantp-ai committed:
Add type annotations to func

dantp-ai committed:
Merge branch 'master' into refactoring/mypy-issues-test-and-examples

dantp-ai committed:
Add type annotations to funcs

  * Make mypy happy and annotate some vars

dantp-ai committed:
Fix policy annotation

dantp-ai committed:
Add type annotations func

dantp-ai committed:
Type policy var to resolve mypy confusion

dantp-ai committed:
Fix mypy issues

  * Use ActionSpaceInfo to type action space attrs
  * Print collector stats using DataclassPPrintMixin
  * Add missing type annotations to funcs

dantp-ai committed:
Check env.spec for none before accessing attrs

dantp-ai committed:
Add type annotation using space info

dantp-ai committed:
Add type annotations to NoopResetEnv's methods and helper funcs

dantp-ai committed:
Use only integers with Generator

  * Generator does not have randint

dantp-ai committed:
Use DataclassPPrintMixin to print collect stats

dantp-ai committed:
Respect mypy typing for vars/args

dantp-ai committed:
Fix many mypy issues

  * type annotations funcs
  * type some vars that mypy asks to
  * Make mypy happy and check for existing attr of ALE-based envs
  * Respect super class methods types

dantp-ai committed:
Rename var to resolve ambiguity for mypy

dantp-ai committed:
Fix mypy issues (see below)

  * Add missing type annotations to func signatures
  * Type ambiguous vars used in if-else conditional blocks to resolve mypy confusion

dantp-ai committed:
Add type annotations to DummyDataset and FiniteEnv

dantp-ai committed:
Fix mypy issues

  * Add missing type annotations to funcs
  * Add typing to some vars for clarify for mypy

dantp-ai committed:
Fix some mypy issues

  * Rename vars to prevent confusion for mypy
  * Add missing function type annotations
  * Type some vars correctly

dantp-ai committed:
Fix many mypy issues related to:

  * Function type annotations
  * Typing some vars to avoid confusion
  * Rewrite some assignments that respect input argument types

dantp-ai committed:
Fix mypy issues:

  * add missing function type annotations
  * use mandatory kw args properly
  * access attrs dataclass with . notation

dantp-ai committed:
ignore mypy check

dantp-ai committed:
Fix mypy issues:

  * add missing function type annotation
  * use dot notation to access dataclass attrs
  * extend some typing
  * adapt output to respect existing type hints

dantp-ai committed:
Add missing type annotations

dantp-ai committed:
Fix some mypy issues:

  * Re-type sleep to float (as it seems that is actually the way it is used)
  * Add some typing to funcs

dantp-ai committed:
Merge branch 'master' into refactoring/mypy-issues-test-and-examples

dantp-ai committed:
Add missing type annotations to funcs

dantp-ai committed:
Fix mypy issues

  * Use ndarray input to step()
  * Use proper type for list elements

dantp-ai committed:
Pass correct typed param env_fns to SubprocVectorEnv

dantp-ai committed:
Fix more mypy issues:

  run_align_norm_obs()
    * Mypy complains about unreachable statements but the logic in the conditional is sound and they are in fact reachable
    * info can be dict or list[dict]. Take into account the latter too.
  * Add type anno to DummyEnv.step()

dantp-ai committed:
Bugfix: Tuple item assignment

dantp-ai committed:
Make mypy happy and use [] instea of . notation

  * mypy does not know that value of keys are Batch

dantp-ai committed:
Merge branch 'master' into refactoring/mypy-issues-test-and-examples

MischaPanch committed:
Typing: extend index type

MischaPanch committed:
Typing in tests: added asserts and cast to remove some mypy errors

dantp-ai committed:
Revert output type to tuple[InfoStats, BasePolicy]

  * Maintain consistency used throughout codebase, using .best_reward

dantp-ai committed:
For mypy: Store obs/obs_next in new var and assert type

dantp-ai committed:
For mypy: store intermediate vars & assert type

dantp-ai committed:
Use pprint_asdict() instead

dantp-ai committed:
Check for none

dantp-ai committed:
Merge branch 'master' into refactoring/mypy-issues-test-and-examples

dantp-ai committed:
Use pprint_asdict to print CollectStats

dantp-ai committed:
Ignore type as mypy doesn't know that it should be wrong here

dantp-ai committed:
Ignore type annotation for step() because env can generate non-scalar rewards:

  * See issue #1080

dantp-ai committed:
Make mypy understand what type weight is

dantp-ai committed:
Make mypy happy and use proper type for adding constant to batch

dantp-ai committed:
Use setattr/getattr because mypy doesn't know

  * of dynamically added attributes
  * ignore ruff checks because ruff would reformat back s.t. mypy will complain (circular)

dantp-ai committed:
Remove unused var

dantp-ai committed:
Use SpaceInfo to determne types action/obs space

dantp-ai committed:
Revert "Use SpaceInfo to determne types action/obs space"

This reverts commit 2e8f3c3e25c161c2b44ff7e207b93f3397604666.

  * SpaceInfo needs to support MultiDiscrete action spaces for the changes in this commit to run successfully

dantp-ai committed:
Cover case when state is dict

  * I made the assumption that state has a key "hidden". Are the keys for an RNN-based state: cell, hidden?

dantp-ai committed:
Set integer default value for batch_size

dantp-ai committed:
Add typing to env methods and use Gym API >v0.26 (with terminated, truncated)

dantp-ai committed:
Treat case when cpu_count() is None

dantp-ai committed:
Fix mypy issues:

  * Add missing type annotations
  * Resolve confusion of unreachable return statement by ignoring that mypy check. If enum has more than one member than mypy won't be confused anymore

dantp-ai committed:
Ignore mypy typing on lines that use old Gym API:

  * mypy doesn't know that Gym version <0.26 has only 4 items (no truncation)

dantp-ai committed:
Make mypy happy and add type to var

dantp-ai committed:
Use space_info to type env spaces

dantp-ai committed:
Use assert instead of cast to check for obs_space/action_space:

  * For situation in which algorithms demand a specific space

dantp-ai committed:
Assert action space

dantp-ai committed:
Use assert as mypy doesn't know that FetchReach env has compute_reward method

dantp-ai committed:
Check for none before comparing mean_rewards to reward_threshold

  * If none, then check is False. Or better to have a default reward_threshold value?

dantp-ai committed:
Assert action space before accessing attributes specific to that space

dantp-ai committed:
Refactor way DQN API is used:

  * Pass c, h, w explicitly
  * Check that obs_space is a tuple of size 3
  * Cast some scalar vars to write type

dantp-ai committed:
Fix FiniteVectorEnv.reset() to satisfy superclass type annotations

dantp-ai committed:
Fix mypy issues for AtariWrappers

  * Use type annotations of Gym API > 0.26 ver
  * Rename some vars to prevent confusion from mypy
  * add missing type annotations
  * add explicit type to some vars to make mypy happy
  * assert that shape obs space is not None before using it

dantp-ai committed:
Make mypy happy and use typing for obs_space_dtype

  * Should we move this code in a utils? or maybe SpaceInfo?

dantp-ai committed:
Assert env.action_space = MultiDiscrete:

  * For MultiDiscrete, shape can never be None (cf. `shape()` in `MultiDiscrete`)

dantp-ai committed:
Extend IndexType to explicitly have list[int]:

  * mypy can't infer that this type is also permitted and so it complains that Collector._reset_state's id is not compatible when used inside the function (e.g. state.empty_(id) when state is Batch)

dantp-ai committed:
Use SpaceInfo to type env obs/action space:

  * Unpack c,h,w. star notation is confusing mypy

dantp-ai committed:
Use SpaceInfo to type env obs/action space:

  * Unpack c,h,w. star notation is confusing mypy

dantp-ai committed:
Assert action space before accessing space-specific attrs

dantp-ai committed:
Add type hints to obs/action spaces:

  * Should `BranchingPolicy` support also `MultiDiscrete` action spaces?

dantp-ai committed:
Add type hints to action/obs spaces:

  * Should `BranchingPolicy` support also `MultiDiscrete` action spaces?

dantp-ai committed:
Minor reformat comments

dantp-ai committed:
Merge branch 'master' into refactoring/mypy-issues-test-and-examples

dantp-ai commented:
@MischaPanch I finished solving the issues. Here is a list with some commits that I think are important to give more attention to:

- https://github.com/thu-ml/tianshou/pull/1077/commits/7c91575a69721270b4fe82c67439d70a8fd28ec3
  - I used the [recommendations from the cheatsheet](https://tianshou.org/en/stable/01_tutorials/07_cheatsheet.html#user-defined-environment-and-different-state-representation).
- https://github.com/thu-ml/tianshou/pull/1077/commits/6f31ac12e935820e0d878d4673ef30aaef41daff
  - ALE and Gym API pre (only done flag) and post version 0.26 (terminated, truncated flags).
- https://github.com/thu-ml/tianshou/pull/1077/commits/f282d814afb3b1a76d4cb97535051de752e9527f
  - I think here one could solve it differently by storing temporary variables and asserting their type, and then continue using dot-notation.
- https://github.com/thu-ml/tianshou/pull/1077/commits/56608dad369dd92cc597396273082687b1a76905
- https://github.com/thu-ml/tianshou/pull/1077/commits/2d6f5f171df0420ee4914e53a0c47d0ff8176a8c
  - mypy does not seem to understand dynamically added attrs to the dataclass.
- https://github.com/thu-ml/tianshou/pull/1077/commits/10352d4a97c19c771696eeefe69e0fb6d03972b0
  - Respect >0.26 Gym API for <0.26 API calls.
- https://github.com/thu-ml/tianshou/pull/1077/commits/2255f6e7c72a6a446ef8d0c2d55c3c79e5ccc292
  - Should `BranchingDQNPolicy` support also `MultiDiscrete` action spaces?
  - I think `tianshou.env.ContinuousToDiscrete` should be renamed to `ContinuousToMultiDiscrete` as it converts action space from `Box` to `MultiDiscrete`. `Discrete` is a special case of `MultiDiscrete`. Also, as it converts only the action space, a better name would be `ContinuousActionSpaceToMultiDiscrete`.

dantp-ai commented:
Now, I will merge the Collector-refactoring PR (https://github.com/thu-ml/tianshou/pull/1063) into my branch before merging this PR to master.

MischaPanch commented on examples/mujoco/plotter.py:
```diff
@@ -3,14 +3,20 @@
+def smooth(
+    y: np.ndarray,
+    radius: int,
+    mode: str = "two_sided",
```
pls use Literal as type, then you can remove the assert

dantp-ai replied:
Yes. it's cleaner.

MischaPanch commented on examples/offline/atari_cql.py:
```diff
@@ -80,16 +82,21 @@ def test_discrete_cql(args: argparse.Namespace = get_args()) -> None:
     torch.manual_seed(args.seed)
     # model
-    net = QRDQN(*args.state_shape, args.action_shape, args.num_quantiles, args.device)
+    net = QRDQN(c, h, w, args.action_shape, args.num_quantiles, args.device)
```
As above. Generally, I feel like the init of such classes should be kwargs only. If you feel like it's too much, can be done in a separate PR

MischaPanch commented on examples/vizdoom/env.py:
```diff
@@ -105,17 +118,27 @@ def step(self, action):
+    seed: int | None = None,
+    training_num: int = 10,
+    test_num: int = 10,
+) -> tuple[Any | Env, Any | ShmemVectorEnv, Any | ShmemVectorEnv]:
```
I think `Any | ...` doesn't make much sense. Why do we need `Any`? If there's no better type, maybe just leave `Any` for now? 

Does this maybe merit a separate issue?

dantp-ai replied:
Thx for spotting it. It should just be `tuple[Env, ShmemVectorEnv, ShmemVectorEnv]`.
I must have used VSCode's double-click to insert the type annotation there.

MischaPanch commented on examples/vizdoom/vizdoom_ppo.py:
```diff
@@ -140,7 +140,8 @@ def test_ppo(args: argparse.Namespace = get_args()) -> None:
         return Categorical(logits=logits)
 
-    policy: PPOPolicy = PPOPolicy(
+    policy: PPOPolicy | ICMPolicy
```
as above

MischaPanch commented on test/base/test_collector.py:
```diff
@@ -210,16 +236,19 @@ def test_collector(gym_reset_kwargs) -> None:
     # test NXEnv
     for obs_type in ["array", "object"]:
-        envs = SubprocVectorEnv([lambda i=x, t=obs_type: NXEnv(i, t) for x in [5, 10, 15, 20]])
+        envs = SubprocVectorEnv([create_env(x, obs_type) for x in [5, 10, 15, 20]])
```
this is less safe than before because of [late binding](https://docs.python-guide.org/writing/gotchas/#late-binding-closures) in functions, a common python gotcha. Pls pass as kwargs instead to bind immediately.

MischaPanch replied:
Done

MischaPanch commented on test/base/test_env_finite.py:
```diff
@@ -62,85 +72,104 @@ def step(self, action):
+        obs: np.ndarray
+        try:
+            obs = np.stack(obs_list)
+        except ValueError:
```
Why is the try-except now necessary?

MischaPanch commented on test/base/test_returns.py:
```diff
@@ -70,7 +71,7 @@ def test_episodic_returns(size=2560) -> None:
         rew=np.array([101, 102, 103.0, 200, 104, 105, 106, 201, 107, 108, 109, 202]),
     )
-    for b in batch:
+    for b in iter(batch):
```
With the newest PR merged, `__iter__` was added to Batch. It should be ok to remove the explicit iter now, right?

dantp-ai replied:
Correct.

MischaPanch commented on test/base/test_stats.py:
```diff
@@ -37,4 +38,8 @@ def test_training_stats_wrapper() -> None:
         wrapped_train_stats.dummy_field = 43
-        assert wrapped_train_stats.wrapped_stats.loss_field == wrapped_train_stats.loss_field == 13
+        assert (
+            getattr(wrapped_train_stats.wrapped_stats, "loss_field")  # noqa: B009
```
no need to use getattr, you can use `type: ignore` or `assert hasattr` to make mypy happy

MischaPanch commented on tianshou/data/batch.py:
```diff
@@ -17,7 +18,14 @@
+    | _SingleIndexType
+    | list[_SingleIndexType]
+    | tuple[_SingleIndexType, ...]
+    | list[int]
```
this shouldn't be necessary, already covered by `list[_SingleIndexType]`

dantp-ai replied:
I can remove it now. There are no issues when not explicitly defining `list[int]`. 

Previously, mypy was complaining because in `Collector._reset_state`, it couldn't infer that `list[_SingleIndexType]` includes `list[int]` too.

MischaPanch replied:
Sounds good

MischaPanch reviewed (changes requested):
Great work @dantp-ai , thanks a  lot for it! It's going to be super useful.

I left a bunch of comments, but apart from them the tests and examples should now be included in the poe task `type-check`, you should modify `pyproject.toml` and `.pre-commit.yaml` for that.

Feel free to give me a call if you want to talk about the comments

MischaPanch commented:
> @MischaPanch Should I first merge the Collector-refactoring PR (#1063) into here, before doing the review of this PR?

@dantp-ai the PR was now merged into master. Could you pls merge master into your branch and resolve the conflicts?

MischaPanch commented:
In fact, I'm going to merge #1032 first and would then resolve the conflicts myself and push to your branch

MischaPanch committed:
Merge branch 'thuml-master' into refactoring/mypy-issues-test-and-examples

# Conflicts:
#	test/base/env.py
#	test/base/test_buffer.py
#	test/base/test_collector.py
#	test/base/test_env.py
#	test/base/test_env_finite.py
#	test/modelbased/test_psrl.py

MischaPanch commented on examples/atari/atari_dqn.py:
```diff
@@ -104,7 +104,8 @@ def test_dqn(args: argparse.Namespace = get_args()) -> None:
     optim = torch.optim.Adam(net.parameters(), lr=args.lr)
     # define policy
-    policy: DQNPolicy = DQNPolicy(
+    policy: DQNPolicy | ICMPolicy
```
Ahh, it's necessary b/c of an open issue in mypy: https://github.com/python/mypy/issues/3737

dantp-ai replied:
I didn't know about the issue in mypy, thx. 

I looked into our issue, whether it's b/c `policy` is defined multiple times with different policies, but mypy wants to type the `policy` even if it is defined once.

Btw, [in the overview tutorial](https://github.com/thu-ml/tianshou/blob/master/docs/02_notebooks/L0_overview.ipynb) `policy` is typed with `BasePolicy`. I think it should be typed with `PPOPolicy` there so that we get all the `PPOPolicy`-specific attrs and methods in the IDE.

MischaPanch replied:
Right, would be better - could you fix it in the tutorial pls?

dantp-ai replied:
Yes.

dantp-ai replied:
Done.

MischaPanch committed:
Post-resolving conflicts in tests and examples

MischaPanch committed:
Pyproject: added stubs, included tests and examples in type-check

MischaPanch commented:
@dantp-ai Done now, conflicts resolved, pipeline should run through. I extended the pyproject. When the remaining issues are addressed, this can be merged.

Pls extend the changelog and copy the relevant parts of the changelog to the PR description (see the last merged PRs for examples on PR descriptions)

MischaPanch committed:
Extended pre-commit type check [skip ci]

MischaPanch committed:
Typo in test

dantp-ai commented on examples/atari/atari_dqn.py:
```diff
@@ -157,6 +158,7 @@ def test_dqn(args: argparse.Namespace = get_args()) -> None:
     log_path = os.path.join(args.logdir, log_name)
 
     # logger
+    logger_factory = LoggerFactoryDefault()
```
A lot of the examples where duplicating code related to logging that already exists in `LoggerFactoryDefault.create_logger()`. [Initially](https://github.com/thu-ml/tianshou/pull/1017#discussion_r1481301565), I created a common `logger_factory`, but as you correctly pointed out that solution is leading to mutable default throughout the code.

The solution here and in other examples is to have a per-example `LoggerFactoryDefault`.

MischaPanch replied:
Ok, that's reasonable

dantp-ai commented on examples/atari/atari_ppo_hl.py:
```diff
@@ -98,10 +98,11 @@ def main(
         .with_epoch_stop_callback(AtariEpochStopCallback(task))
     )
     if icm_lr_scale > 0:
+        hidden_sizes = [hidden_sizes] if isinstance(hidden_sizes, int) else hidden_sizes
```
Yes, a sequence for `hidden_sizes` would be way easier to work with and it includes the case with one layer. So no need to distinguish it with a special type (if that is the reason of allowing an `int` now).

MischaPanch replied:
I don't think there's a good reason for this at all. Should it be done in a separate PR, or is the change small enough that you want to do it here?

dantp-ai replied:
It can be done here. There are only two examples using `ActorFactoryAtariDQN`.

dantp-ai replied:
`ActorFactoryAtariDQN` currently names it `hidden_size`. Better naming for sequence is `hidden_sizes`.

dantp-ai commented on examples/atari/atari_network.py:
```diff
@@ -251,12 +252,25 @@ def __init__(
             device=device,
             features_only=self.features_only,
-            output_dim=self.hidden_size,
+            output_dim=self.hidden_size
```
> Have you tested that atari examples run through?
Yes.

> why did it work before this?
All the examples which use `ActorFactoryAtariDQN` are passing an int for `hidden_size`.

> this is a logical change,
Yes, so I assume that if it's a sequence of hidden sizes, then the last value should be the last layer which coincides with the output dimension.

dantp-ai replied:
If we go with the change in `ActorFactoryAtariDQN` of `hidden_size: int | Sequence[int]` to `hidden_sizes: Sequence[int]` then the logic above to `output_dim` needs to be considered.

dantp-ai commented on examples/box2d/bipedal_bdq.py:
```diff
@@ -125,7 +128,9 @@ def save_best_fn(policy: BasePolicy) -> None:
 
     def stop_fn(mean_rewards: float) -> bool:
-        return mean_rewards >= getattr(env.spec.reward_threshold)
+        if env.spec and env.spec.reward_threshold:
```
Yes true, it would have led to an error, but `stop_fn` is not being used in this example. For unknown reasons it's commented out when passed to the `OffpolicyTrainer` [here](https://github.com/thu-ml/tianshou/pull/1077/files/d01420c714d0a2bac82cd72027b8c7b3cca265ba#diff-3792a1a111e15c573f79326f5922fa97aad9f2022aa90d13a088ffada9b78434R153).

MischaPanch replied:
Wanna fix this? :)

I guess someone accidentally broke it and then just commented it out

dantp-ai replied:
The `epoch` was set by default to 1000. I guess (and hope) that the agent can reach `reward_threshold` way sooner (like 10-20 epochs as in other examples). I am running now the script to determine what the lowest epoch is s.t. `stop_fn` is true.

dantp-ai commented on examples/mujoco/mujoco_reinforce_hl.py:
```diff
@@ -29,7 +29,7 @@ def main(
     step_per_collect: int = 2048,
     repeat_per_collect: int = 1,
-    batch_size: int | None = None,
+    batch_size: int = 16,
```
I introduced here an `int` by default because `None` would not be compatible with `batch_size` in `SampleConfig` which is always an `int`. 

If we allow `None`, there is an extra (assert) check to be done to make mypy happy.

Should we rather change the default value from 16 to 64 as it is by default in `SampleConfig` (used a couple of lines below) ?

MischaPanch replied:
There is no good default value for batch size. None has a certain semantics now which was introduced with a PR not long ago. I guess sampling config should accept None and the docstring should be extended. If you want, I can do that

dantp-ai replied:
> If you want, I can do that
OK.

MischaPanch replied:
Done

dantp-ai committed:
Remove type in IndexType for batch

MischaPanch commented on examples/mujoco/fetch_her_ddpg.py:
```diff
@@ -96,33 +101,37 @@ def test_ddpg(args: argparse.Namespace = get_args()) -> None:
     env, train_envs, test_envs = make_fetch_env(args.task, args.training_num, args.test_num)
+    # The method HER works with goal-based environments
+    assert isinstance(env.observation_space, gym.spaces.Dict)
+    assert hasattr(env, "compute_reward")
```
Pls add this explanation as content of a ValueError

dantp-ai replied:
Done.

dantp-ai committed:
Use assert hasattr instead of getattr

dantp-ai committed:
Remove iter since Batch already implements __iter__

dantp-ai committed:
Use Literal instead of asserting members of list

dantp-ai committed:
Use stop_fn for running this example

  * Change default num epochs 1000 -> 25

dantp-ai committed:
Use ValueError to inform user about what type of env is supported.

dantp-ai committed:
Use more specific type hint for policy to get access to policy-specific attrs and methods

dantp-ai committed:
Refactor type annotation make_vizdoom_env

dantp-ai committed:
Use kw-args for better readability

dantp-ai committed:
Use os.path.join

dantp-ai committed:
Remove if and assert hasattr beforehand

dantp-ai committed:
Return non-empty dict when reset

dantp-ai committed:
Refactor ActorFactoryAtariDQN hidden_size semantics and output_dim of APIs using it.

  * `hidden_size` is replaced with `output_dim_added_layer` to better reflect its purpose, which is to add an output layer to DQN body net or not (based on whether `features_only`)

dantp-ai committed:
Refactor type annotations of scale_obs:

  * it operates on Nets not Modules

dantp-ai committed:
Simplify checks of obs_shape for atari envs

dantp-ai committed:
Use kw for input arguments to QRDQN

MischaPanch committed:
Made NetBase generic (explanation below), removed **kwargs from forward

The generic was needed because of the type of "state", representing the hidden state

The only place where it is ever used is in recurrent version of DQN

MischaPanch committed:
SamplingConfig: support for batch_size=None

MischaPanch committed:
Changelog [skip ci]

MischaPanch committed:
Merge branch 'thuml-master' into refactoring/mypy-issues-test-and-examples

# Conflicts:
#	CHANGELOG.md

MischaPanch committed:
Changelog [skip ci]

maxhuettenrauch commented on examples/atari/atari_wrapper.py:
```diff
@@ -191,17 +195,24 @@ class WarpFrame(gym.ObservationWrapper):
+            obs_space_dtype = np.integer
+        elif np.issubdtype(type(obs_space.dtype), np.floating):
+            obs_space_dtype = np.floating
+        assert isinstance(obs_space, gym.spaces.Box)
```
Hi @dantp-ai ! Could you let me know what was the reason for this change? It leads to a broken DQN example on Atari for me due to `obs_space_dtype` being undefined. The type check only works if I check `np.issubdtype(obs_space.dtype, np.integer)`, otherwise it will be False...

dantp-ai replied:
Let me take a look how we can write this better.

The reason for the modification was a type mismatch previously between what `Box` allows for `dtype` compared to the more general type of `Space` for `dtype:

```
examples/atari/atari_wrapper.py: note: In member "__init__" of class "WarpFrame":
examples/atari/atari_wrapper.py:212: error: Argument "dtype" to "Box" has incompatible type
"dtype[Any] | dtype[void] | None"; expected "type[floating[Any]] | type[integer[Any]]"  [arg-type]
                dtype=env.observation_space.dtype,
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

maxhuettenrauch replied:
... and also a numpy DeprecationWarning for setting np.integer as dtype.

dantp-ai replied:
I found a solution:

```
assert isinstance(obs_space, gym.spaces.Box)

dtype: type[np.integer] | type[np.floating]
if np.issubdtype(obs_space.dtype, np.floating):
    dtype = np.float32
else:
    dtype = np.int32

self.observation_space = gym.spaces.Box(
            low=np.min(obs_space.low),
            high=np.max(obs_space.high),
            shape=(self.size, self.size),
            dtype=dtype,
        )
```

One concern I have is that there could be other suitable floating-point types (e.g., `np.float64`) based on the system.

maxhuettenrauch replied:
Yeah, I also feel that we shouldn't overwrite the dtype of the original observation space. What do you think @MischaPanch ?

maxhuettenrauch replied:
@MischaPanch , this is probably the root for issue #1111

dantp-ai replied:
Yes. Should I go ahead and ignore the type issue here as it becomes so complex to implement it (cf. above one possible solution)?

MischaPanch replied:
Sorry, it had escaped my attention. How about just replacing the assert by cast?

dantp-ai replied:
Unfortunately casting the `obs_space` is weaker than the assert and mypy will complain that some attrs don't exist because it won't assume the space to be `Box`.

MischaPanch replied:
I'll check and will push the fix directly to master. We need to find some way of testing the examples, I'm going to contact some people from AMD and ask them to donate us CI workers

dantp-ai replied:
Ok, mind that this dtype issue is in `WarpFrame` and `FrameStack`.

MischaPanch replied:
That was just a bug - the `type` too much in the isinstance check. I factored it out into a function and fixed it :)
